homeassistant/components/sonos/sensor.py

```diff
@@ -7,22 +7,36 @@
     ENTITY_CATEGORY_DIAGNOSTIC,
     PERCENTAGE,
 )
+from homeassistant.core import callback
 from homeassistant.helpers.dispatcher import async_dispatcher_connect
 
-from .const import SONOS_CREATE_BATTERY
+from .const import SONOS_CREATE_AUDIO_FORMAT_SENSOR, SONOS_CREATE_BATTERY
 from .entity import SonosEntity
 from .speaker import SonosSpeaker
 
 
 async def async_setup_entry(hass, config_entry, async_add_entities):
     """Set up Sonos from a config entry."""
 
-    async def _async_create_entity(speaker: SonosSpeaker) -> None:
+    @callback
+    def _async_create_audio_format_entity(speaker: SonosSpeaker) -> None:
+        entity = SonosAudioInputFormatSensorEntity(speaker)
+        async_add_entities([entity])
+
+    @callback
+    def _async_create_battery_sensor(speaker: SonosSpeaker) -> None:
         entity = SonosBatteryEntity(speaker)
         async_add_entities([entity])
 
     config_entry.async_on_unload(
-        async_dispatcher_connect(hass, SONOS_CREATE_BATTERY, _async_create_entity)
+        async_dispatcher_connect(
+            hass, SONOS_CREATE_AUDIO_FORMAT_SENSOR, _async_create_audio_format_entity
+        )
+    )
+    config_entry.async_on_unload(
+        async_dispatcher_connect(
+            hass, SONOS_CREATE_BATTERY, _async_create_battery_sensor
+        )
```

Comment:
So, this property needs polling (just like the battery), should both of this share the same code path? Is it possible to share some code among these?

Reply:
The codepaths are totally different. The battery info is polled, but from a central polling handler in `speaker.py` (not the built-in entity polling). The battery polling is also intentionally throttled to avoid hitting the battery-operated devices too frequently.

Reply:
Ok, I have no clue about the internals, I just recalled hearing/reading that the battery is a pollable entity and I was thinking this would be alike. So just forget this :-)